src/gp_init.c

```diff
@@ -376,47 +376,101 @@ int init_event_fini(struct gssproxy_ctx *gpctx)
     return 0;
 }
 
-void init_proc_nfsd(struct gp_config *cfg)
+static int init_proc_nfsd_once(void)
 {
     char buf[] = "1";
-    bool enabled = false;
     int fd, ret;
-    static int poked = 0;
-
-    /* check first if any service enabled kernel support */
-    for (int i = 0; i < cfg->num_svcs; i++) {
-        if (cfg->svcs[i]->kernel_nfsd) {
-            enabled = true;
-            break;
-        }
-    }
+    static bool poked = false;
+    static bool warned_once = false;
 
-    if (!enabled || poked) {
-        return;
-    }
+    if (poked)
+        return 0;
 
     fd = open(LINUX_PROC_USE_GSS_PROXY_FILE, O_RDWR);
     if (fd == -1) {
         ret = errno;
-        GPDEBUG("Kernel doesn't support GSS-Proxy (can't open %s: %d (%s))\n",
-                LINUX_PROC_USE_GSS_PROXY_FILE, ret, gp_strerror(ret));
-        goto fail;
+        if (!warned_once) {
+            GPDEBUG("Kernel doesn't support GSS-Proxy "
+                    "(can't open %s: %d (%s))\n",
+                    LINUX_PROC_USE_GSS_PROXY_FILE, ret, gp_strerror(ret));
+            warned_once = true;
+        }
+        goto out;
     }
 
     ret = write(fd, buf, 1);
     if (ret != 1) {
         ret = errno;
         GPDEBUG("Failed to write to %s: %d (%s)\n",
                 LINUX_PROC_USE_GSS_PROXY_FILE, ret, gp_strerror(ret));
-        close(fd);
-        goto fail;
+        goto out_close;
     }
 
-    poked = 1;
+    GPDEBUG("Kernel GSS-Proxy support enabled\n");
+    poked = true;
+    ret = 0;
+out_close:
     close(fd);
-    return;
-fail:
-    GPDEBUG("Problem with kernel communication!  NFS server will not work\n");
+out:
+    return ret;
+}
+
+static void delayed_proc_nfsd(verto_ctx *vctx UNUSED, verto_ev *ev)
+{
+    struct gssproxy_ctx *gpctx;
+    int ret;
+
+    gpctx = verto_get_private(ev);
+
+    ret = init_proc_nfsd_once();
+    if (ret == 0) {
+        verto_del(gpctx->retry_proc_ev);
+        gpctx->retry_proc_ev = NULL;
+    }
+}
+
+int init_proc_nfsd(struct gssproxy_ctx *gpctx)
+{
+    bool enabled = false;
+    int ret;
+
+    /* check first if any service enabled kernel support */
+    for (int i = 0; i < gpctx->config->num_svcs; i++) {
+        if (gpctx->config->svcs[i]->kernel_nfsd) {
+            enabled = true;
+            break;
+        }
+    }
+
+    if (!enabled) {
+        goto out;
+    }
+
+    ret = init_proc_nfsd_once();
+    if (ret == 0) {
+        goto out;
+    }
+
+    /* failure, but the auth_rpcgss module might not be loaded yet */
+    if (!gpctx->retry_proc_ev) {
+        gpctx->retry_proc_ev = verto_add_timeout(gpctx->vctx,
+                                                 VERTO_EV_FLAG_PERSIST,
```

Comment:
This means there will be cases where this may fire over and over forever if init_proc_nfsd_once() always fail ... is this intentional?

Reply:
Yes, it'll fire every 10 seconds forever....*but* it will only do so if:
a) the user has a nfs-server configuration for gssproxy; and
b) /proc/net/rpc/use-gss-proxy doesn't exist (i.e. the auth_rpcgss module isn't loaded)

I do realize that the default `gssproxy` installation includes `/etc/gssproxy/24-nfs-server.conf` though....I still think a wakeup every 10s is acceptable?

Reply:
I would think maybe we should have a very simple backoff.
Store the delay in another static variable, start with one second delay and double it each time we call the function until it reaches 1024 seconds (slightly more than 15 minutes), and then keep it there?

Reply:
I think that has the potential of being quite confusing. The admin installs `gssproxy`...starts it, then installs `nfs-utils` a couple of minutes later....it doesn't work straight away but 15 minutes later it magically starts working?

`libverto` already has a 60s wakeup interval even when gssproxy isn't doing anything....maybe use that as the upper limit?

```
root@qtest1:~# ps ax | grep gssproxy
    961 ?        Ssl    0:00 /usr/sbin/gssproxy -D
root@qtest1:~# strace -t -p 961 (that's gssproxy)
strace: Process 961 attached
19:08:09 epoll_wait(8, [], 64, 59743)   = 0
19:09:09 epoll_wait(8, [], 64, 59743)   = 0
19:10:09 epoll_wait(8, [], 64, 59743)   = 0
19:11:09 epoll_wait(8, [], 64, 59743)   = 0
19:12:09 epoll_wait(8, [], 64, 59743)   = 0
19:13:08 epoll_wait(8, [], 64, 59743)   = 0
19:14:08 epoll_wait(8, [], 64, 59743)   = 0
```

Though I really don't think a 10s interval is a real problem either....it's basically 1-2 system calls per 10s....won't even show up in `top` on the original Raspberry Pi :)